tests/apps/test_graph_max_clique.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Unit tests for strawberryfields.apps.graph.max_clique
+"""
+# pylint: disable=no-self-use,unused-argument
+import pytest
+import networkx as nx
+import numpy as np
+
+from strawberryfields.apps.graph import max_clique
+from strawberryfields.apps.graph import resize
+
+
+def patch_random_choice(x):
+    """Dummy function for ``np.random.choice`` to make output deterministic. This dummy function
+    just returns the element of ``x`` specified by ``element``."""
+    if isinstance(x, int):  # np.random.choice can accept an int or 1D array-like input
+        return 0
+
+    return x[0]
+
+
+def patch_clique_resize(clique, graph, node_select):
+    """Dummy function for ``clique_grow`` and ``clique_swap`` to make output unchanged."""
+    return clique
+
+
+class TestLocalSearch:
+    """Tests for the function ``max_clique.local_search``"""
+
+    def test_bad_iterations(self):
+        """Test if function raises a ``ValueError`` when a non-positive number of iterations is
+        specified"""
+        with pytest.raises(ValueError, match="Number of iterations must be a positive int"):
+            max_clique.local_search(clique=[0, 1, 2, 3], graph=nx.empty_graph(5), iterations=-1)
+
+    def test_max_iterations(self, monkeypatch):
+        """Test if function stops after 5 iterations despite not being in a dead end (i.e., when
+        ``grow != swap``). This is achieved by monkeypatching the ``np.random.choice`` call in
+        ``clique_grow`` and ``clique_swap`` so that for a 5-node wheel graph starting as a [0,
+        1] node clique, the algorithm simply oscillates between [0, 1, 2] and [0, 2, 3] for each
+        iteration of growth & swap. For odd iterations, we get [0, 2, 3]."""
+
+        graph = nx.wheel_graph(5)
+        clique = [0, 1]
+
+        with monkeypatch.context() as m:
+            m.setattr(np.random, "choice", patch_random_choice)
+            result = max_clique.local_search(clique, graph, iterations=5)
+
+        assert result == [0, 2, 3]
+
+    def test_dead_end(self, monkeypatch):

```

Comment:
I'm not sure this test is doing anything more than testing whether the line `if grow == swap or iterations == 0:` is working properly. Is that enough?

Reply:
> Although I guess tests are not just there to test existing code but to check if we make changes to the code that there are no errors.

Yes :100:. I agree with keeping this test in for this reason. Tests should treat functions as a black box and not care about internal implementation precisely because it can change.